Cylc-using sites: interactive world map & nested list

sadielbartholomew opened the pull request:
Closes #17, if there is consensus that the idea is a good one?

*I created the code for the interactive world map of Cyc-using sites (based on the list on the [current site 'Cylc users' page](https://cylc.github.io/users.html)) for our web-based slides for the upcoming RSEConUK '19, & I thought it might be a good addition to the website. If you don't like it, I am happy to close this, however.*

Implementation details can be tweaked as desired. Styling etc. can be easily customised, notably the 'watercolour' background map theme which I really like but which others may not. There are some improvements I would want to make eventually, such as preventing the overlapping of labels for clusters of sites e.g. in the UK sites, when viewing zoomed out to close to a whole-world view, but I do not have time for those now.

Additionally, while addressing the user-base page, I am suggesting a small improvement (as I see it at least) where the 'known users' listing is made nested to be organised by continent & further by country, to organise the sites in a way that both:
* is easier to comprehend, & to search for an item within;
* emphasises the international nature of Cylc usage (to complement the world map which has the same aim).

### Note

I have tested that the ``sites-world-map.html`` map opens & renders in the browser. It should look like so, & you should be able to explore (zoom, move around) the map too:

![world_map_scr](https://user-images.githubusercontent.com/30274190/63962360-9d9b2980-ca8a-11e9-984c-f8532f520bae.png)

![world_map_scr2](https://user-images.githubusercontent.com/30274190/63962537-f5d22b80-ca8a-11e9-8895-e9f00f141170.png)

But I have not tested that the site after the change here builds to render everything correctly, as I believe I need Jekyll to test a build locally, & [in turn need a Ruby environment](https://jekyllrb.com/docs/), which I am not sure I can get at work. Shortly I will see if I can build to test the PR on my personal laptop.

Logically, though, the content should build & display the same '``user.html`` page, the only differences being the world map displayed at the top of the page contents, followed by the list, now nested.

sadielbartholomew committed:
Cylc-using sites: add interactive world map & organise the list

kinow commented:
Screenshot of the users page before:

![image](https://user-images.githubusercontent.com/304786/63985376-8a886800-cb23-11e9-9967-6b92b0c4583c.png)

And after this PR:

![image](https://user-images.githubusercontent.com/304786/63985344-67f64f00-cb23-11e9-8af5-8565e866745d.png)

I think the map needs further work perhaps. It occupies a large area, and zooming out (which happened by accident while scrolling) causes it to render a lot of nodes, and also get a weird sizing issue.

![image](https://user-images.githubusercontent.com/304786/63985418-c7ecf580-cb23-11e9-837e-14bfe905859d.png)

kinow commented on _includes/sites-world-map.html:
```diff
@@ -0,0 +1,134 @@
+      "NOAA Center for Weather and Climate Prediction (NCWCP)": [38.972, -76.925],
+      "Euro-Mediterranean Center on Climate Change (CMCC)": [40.35225, 18.17165],
+      "Barcelona Supercomputing Center (BSC)": [41.389, 2.116],
+      "National Center for Atmospheric Reserach (NCAR)": [39.97788, -105.27486],
```
s/Reserach/Research

kinow replied:
Old typo too

sadielbartholomew replied:
Yes, I didn't change any text I just transposed it for the nested list. But good spot, I'll amend any typos found now.

kinow commented on users.md:
```diff
@@ -1,40 +1,59 @@
+    * [National Centre for Medium Range Weather
+      Forecasting](http://www.ncmrwf.gov.in/)
+  * Korea:
+  * [Korean Meteorological Administration (KMA)](http://web.kma.go.kr/)
```
Missing space here to make it a sub-item

sadielbartholomew replied:
True, good spot.

kinow commented on users.md:
```diff
@@ -1,40 +1,59 @@
+    * [NRL Marine Meteorology Division](http://www.nrlmry.navy.mil)
+    * [557th Weather Wing](https://www.557weatherwing.af.mil)
+    * [Geophysical Fluid Dynamics Laboratory](http://www.gfdl.noaa.gov/)
+    * [National Center for Atmospheric Reserach (NCAR)](https://ncar.ucar.edu)
```
s/Reserach/Research

kinow replied:
Though this is an old typo and could be fixed separately

sadielbartholomew replied:
Old, yes, but I'll fix it here anyway.

kinow commented on users.md:
```diff
@@ -1,40 +1,59 @@
-  * [National Centre for Medium Range Weather
-      Forecasting](http://www.ncmrwf.gov.in/) - India
-  * [Korean Meteorological Administration (KMA)](http://web.kma.go.kr/) - Korea
+* Oceania:
```
Should we sort these lists alphabetically?

sadielbartholomew replied:
Just to check, you mean within the sub-lists? Or do you mean alphabetise the overall list, as an alternative to the overall idea of nesting the list by continent & country? Either way makes sense, & I am happy to change as we see best.

kinow replied:
Just within the sub-lists, and maybe the continents as well. I liked the separation per continent, gave me a much strong impression of the presence of Cylc worldwide :tada:

kinow commented on _includes/sites-world-map.html:
```diff
@@ -0,0 +1,134 @@
+    // Lat-lon decimal degrees coordinates (to ideally 5 s.f.) for major...
+    // ... known Cylc-using sites, with keys as neat names for labelling:
+    var cylcSiteLatLonCoors = {
+      "Met Office": [50.72762, -3.47564],
```
I think it would be nice if we could have this list of sites as a variable like

```js
const cylcSites = [
{
name: "Met Office",
latlon: [50.72762, -3.47564],
continent: 'Europe'
},
...
]
```

or organize by continents if easier. And then derive from the list both the map, as well as the list below. So maintaining it would be easier, as we could update just one place in a file.

sadielbartholomew replied:
Good idea! For extensions, I was thinking we could display extra information via e.g. the node colours, for example to distinguish sites where Cylc is developed as well as used i.e. MO, NIWA (& BoM?), or UM partner sites, & this as a data structure could make that much easier to add that in, by setting a relevant key to a Boolean & processing that.

When I get around to revising this, I will get this data structure set up.

hjoliver commented:
> notably the 'watercolour' background map theme which I really like but which others may not.

I quite like that (I was a keen amateur watercolour artist before cylc and kids took over my entire life).

hjoliver commented:
@sadielbartholomew - I generally agree with @kinow's comments - but this is a nice start.  The page will certainly look pretty slick with such a map on it, if the zoom and overlap problems can be solved.

sadielbartholomew commented:
> I was a keen amateur watercolour artist before cylc and kids took over my entire life

Ah I remember you saying similar when you visited Exeter last year, that's very cool! I hope you can get some time to restart on the art if you wish to, when you find (eventually... with the kids growing up perhaps?) yourself with more time.

sadielbartholomew commented:
> I think the map needs further work perhaps. It occupies a large area, and zooming out (which happened by accident while scrolling) causes it to render a lot of nodes, and also get a weird sizing issue.

>The page will certainly look pretty slick with such a map on it, if the zoom and overlap problems can be solved.

Thanks for the feedback @kinow & @hjoliver.

I certainly agree the formatting needs amending, now I have seen the screenshots (thanks further @kinow, they are really helpful); if you hadn't spotted from my 'note' at the end of the opening comment, I hadn't actually so far built the page to view it with the map (which I had viewed to test in the browser independently) added in, since it required Ruby.

It should be easy to fix the main issues of dimensions & zoom, I will just change the ``height`` & ``width`` specs, & prevent the map from wrapping continuously horizontally with a ``wrapX = false`` addition. The labelling overlapping is not so easy, but it may look okay anyway once the previous issues are addressed. I'll see & get back to you.

kinow commented:
>I hadn't actually so far built the page to view it with the map (which I had viewed to test in the browser independently) added in, since it required Ruby.

Aahhh! I re-read it now, and it makes sense. Yesterday when reading it for some reason I got the idea that you were warning others about the need of Ruby/Jekyll :smile: If you don't want to mess with your environment installing jekyll and Ruby, you can use GitHub to preview your work.

You would have to enable GitHub Pages in your cylc.github.io fork, push your branch to the GitHub Pages configured branch (watching out to push to the right `remote`), and in seconds it should be built and you will have the site preview at https://sadielbartholomew.github.io/cylc.github.io/index.html

Annoying as you would have to push and wait to see the change, but just in case it helps :+1: 

Thanks Sadie!
Bruno

sadielbartholomew commented:
> you can use GitHub to preview your work

Ah, nice! I did not realise that could be done. Thank you greatly for the tip, it will be much easier to get something that looks right when I can see it built 😄 I have since thought I could also hack it to view by copying all the relevant HTML, i.e. the ``user.md`` page built into the corresponding HTML, with the map HTML added in the place it should be from the ``include``, but this seems much simpler & reliable 🙌

sadielbartholomew committed:
Users list: correct old typos & alphabetise nested items

sadielbartholomew committed:
Improve labelling: use acronym at low zoom, mark dev & UM sites, etc.

sadielbartholomew committed:
Syntax improvements after recalling JS objects are not Python dicts

sadielbartholomew committed:
Change display of map within page, & add brief text to describe it

sadielbartholomew commented:
This is **now ready for re-review**. Many thanks to @kinow for his tip, allowing me to see the end result so I could tweak it! In terms of changes, I have addressed the feedback so far, notably by:

* adjusting the page dimensions & overall display:
    * preventing horizontal wrap, so only one bounded view of the entire world can be seen;
    * setting a minimum zoom level so that it is not possible to get the ugly view with all the labels dominating the view;
    * moving the overall map into the ``page`` class so it is displayed with the same width applied as the other page contents.
It now looks like so (at default, & now maximum possible, zoom out):
![worldmap_scr6](https://user-images.githubusercontent.com/30274190/64082497-db58b600-cd07-11e9-86fa-fbd26434a378.png)
* coordinating the labels (note the labelling, especially over Western Europe, is still not the neatest, but it was *really* difficult & fiddly (& time-consuming) to get all the text to not overlap at full-world zoom-out, even with by doing all of the below, so this is the best I can do for now, as I need to move on):
    1. making the label text show the acronym of a site (as most have a dedicated one) instead of the full name, when the zoom level is below a cutoff; the short-hand version is as shown in the screenshot above, where for higher zoom you get e.g:
        ![worldmap_scr8](https://user-images.githubusercontent.com/30274190/64082608-8158f000-cd09-11e9-83df-069f01c291d7.png)
    2. manually overriding the offsets on the labels to force them to certain positions to fit required text in alongside the surrounding sites (it's a bit hacky, but it is best solution I can find, & at least means full control);
    3. grouping together labels for some co-located sites (ESiWACE & DKRZ, noting that the Max Planck Institute is very close but not on the exact same site), e.g:
        ![worldmap_scr9](https://user-images.githubusercontent.com/30274190/64082700-eb25c980-cd0a-11e9-99bb-582558133a01.png)
    4. adding some colour to help distinguish text, but also to mark the main sites where Cylc is developed & UM Partner sites (though if we don't want to include that info, I can remove it).
* changing the data structure holding the data on Cylc sites according to @kinow's suggestion, though I have not yet included info on county & continent & made it so that the listing is auto-generated from that (that is quite a bit more work, & if we merge this I can open that as a new Issue & address it later).

This (so far) has been a great exercise in JS (I can't say I would have taken this on knowing how much I would need to do to get something decent, but oh well, learning is always good)! Some particularly nice things I learnt on the way were:
* [object spread to merge objects](https://github.com/cylc/cylc.github.io/blob/76ce876d31c69c5d44391fa465627e9e6891bbcf/_includes/sites-world-map.html#L189-L192);
* the ``map()` & ``reduce()`` methods;
* [basic JS regular expression patterns](https://github.com/cylc/cylc.github.io/blob/76ce876d31c69c5d44391fa465627e9e6891bbcf/_includes/sites-world-map.html#L248).

Also I noticed I was writing JS in a very Python-like way (see 5d1116527b16a4ddf51492dd449daf9f1034460e)! I hope I can ease into a "JS mode" in future :smile:

hjoliver reviewed (approved):
Nice work @sadielbartholomew :+1: I'd be happy for this to go in as-is, perhaps with some minor improvements in the future - e.g.:
- a reset button to restore the original view after a bunch of scrolling and zooming;
- maybe code to prevent scrolling off the boundaries of the map.

kinow commented:
> if we merge this I can open that as a new Issue & address it later).

+1

>This (so far) has been a great exercise in JS (I can't say I would have taken this on knowing how much I would need to do to get something decent, but oh well, learning is always good)! 

That's the spirit! Lots of of JS in Cylc UI if you liked it :slightly_smiling_face: 

>Also I noticed I was writing JS in a very Python-like way (see 5d11165)! I hope I can ease into a "JS mode" in future

Haha, nice one. I find myself "tabbing" my JS code sometimes after switching Cylc Flow to Cylc UI :smile:

kinow commented on _includes/sites-world-map.html:
```diff
@@ -0,0 +1,320 @@
+        new ol.layer.Tile({
+          source: new ol.source.Stamen({
+            layer: 'watercolor',
+            wrapX: false, // disable continuously-wrapped horiztonal map.
```
nit pick and can be fixed later, so do not worry about fixing unless changing this file again s/horiztonal/horizontal

hjoliver replied:
Actually can we turn continuous horizontal scrolling back on?  That would largely solve the "scroll off the edge" problem (which affects NZ quite badly!)

hjoliver replied:
And also, we could start with NZ in the middle :grin:

kinow commented on _includes/sites-world-map.html:
```diff
@@ -0,0 +1,320 @@
+      view: new ol.View({
+        center: ol.proj.fromLonLat([30, 0].reverse()),
+        zoom: 2,
+        minZoom: 2 // prevent zoom out whereby labels ugily overwhelm view
```
s/ugily/uglily ?

hjoliver replied:
and then, s/uglily/ugly ?  :rofl:

sadielbartholomew replied:
Hmm, maybe I will change the word to avoid us trying to work out how to phrase it! :smile:

kinow commented on users.md:
```diff
@@ -1,41 +1,69 @@
+      Forecasting](http://www.ncmrwf.gov.in/)
+  * Korea:
+    * [Korean Meteorological Administration (KMA)](http://web.kma.go.kr/)
+  * Phillippines:
```
s/Phillippines/Philippines

sadielbartholomew replied:
Very well spotted!

kinow commented on users.md:
```diff
@@ -1,41 +1,69 @@
+  * Korea:
+    * [Korean Meteorological Administration (KMA)](http://web.kma.go.kr/)
+  * Phillippines:
+    * [Phillippine Atmospheric, Geophysical, and Astronomical Services
```
s/Phillippine/Philippine

kinow replied:
These are probably old typos, we can fix them in a future PR (maybe when creating the list from the JS data structure :)

hjoliver replied:
Yes probably my fault.  @sadielbartholomew might as well fix them though, now that they've been spotted.

kinow commented on _includes/sites-world-map.html:
```diff
@@ -0,0 +1,320 @@
+        })
+      ],
+      view: new ol.View({
+        center: ol.proj.fromLonLat([30, 0].reverse()),
```
How about 

```
center: ol.proj.fromLonLat([25, 25].reverse()),
```

?

This way we will make sure New Zealand is on the map by default (so our site doesn't go to https://www.reddit.com/r/MapsWithoutNZ/). I think it displays all the sites this way by default?

![image](https://user-images.githubusercontent.com/304786/64086632-83649280-cd8d-11e9-97ee-a9b8734f749c.png)

Except for this, ready to merge :tada: Great work @sadielbartholomew !! :clap:

sadielbartholomew replied:
Hmmm, it was definitely on the map when I viewed it, since I went to check that for you NZ based folks, but maybe it depends on the screen resolution. Though it is stuck right at the side, which is not ideal, so I will change to make it the central point of the map as displayed initially (before user interaction).

kinow reviewed (changes requested):
Don't mind the typos @sadielbartholomew , but would be nice to move the center to 25, 25, so NZ is displayed initially too, IMO

kinow commented:
Oh, and the red dots are sites with developers? Should we make BoM red as well?

sadielbartholomew commented:
Thanks both for the new feedback, which along with the previous is really useful for the PR of course but also for refining the map for use in the RESConUK presentation slides (we'll have the fanciest slides this side of town :sunglasses:!).

> I'd be happy for this to go in as-is, perhaps with some minor improvements in the future - e.g. ...

I agree there are certainly a number of improvements that can be made, but since I did this predominantly for the presentation materials which are deadlined, I can't work on this any more, as I have set Cylc dev priorities assigned (the auth & encryption, etc.) :smile: Good suggestions, & once this is merged I will write these up, along with a few of my own additional points, in an Issue.

I'll add another commit in a moment to address the new comments.

sadielbartholomew committed:
Respond to feedback, esp. to put NZ back on (the centre of) the map

sadielbartholomew commented:
Feedback addressed! New Zealand has taken it's rightful place nearer the centre, but not in the very centre, as originally intended, since that would cut the European cluster off, so I made it essentially horizontally central such that the USA & European clusters were not forced too close to the edges:

![worldmap_scr15](https://user-images.githubusercontent.com/30274190/64129103-f54ed300-cdb1-11e9-8fa8-ece990bdc2fd.png)

(*Edit*: oops, the screenshot above was before I addressed the BoM as a dev site comment; it will be red instead for the built branch here).

Also, continuous horizontal scrolling has been re-enabled, as per Hilary's request.

Kindly re-review when you get a chance, @kinow. Thanks again.

kinow commented:
We got a nice improvement to our users page! Thanks Sadie!

hjoliver commented:
Brilliant - thanks @sadielbartholomew!
